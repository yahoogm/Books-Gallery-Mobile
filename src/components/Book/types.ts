import {VolumeInfo} from '../../types/types';

type DetailBookType = {
  volumeInfo: VolumeInfo;
};

export type BookProps = {
  uriDetailBook: string;
  authorsName: string[] | string;
  categories: string[] | string;
  detailBook: DetailBookType;
  setShowBook: React.Dispatch<React.SetStateAction<boolean>>;
  refReadBook: React.RefObject<any>;
};

export type BookContainerProps = {
  setShowBook: React.Dispatch<React.SetStateAction<boolean>>;
  refReadBook: React.RefObject<any>;
};
