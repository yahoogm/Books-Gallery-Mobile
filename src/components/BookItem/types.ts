export type BookItemProps = {
  detailBook: string | number | string[];
  label: string;
};
