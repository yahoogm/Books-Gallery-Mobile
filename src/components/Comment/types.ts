import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {BookItem, Comment, RootStackParamList} from '../../types/types';
import {ObjectSchema} from 'yup';

export type FormCommentValues = {
  comment: string;
};

export type AddCommentValues = {
  userName: string;
  bookId: string;
  profilePic: string;
  id: string;
  ulasan: string;
  userId: string;
  createdAt: string;
  updatedAt: string;
};

// type for handling navigation to login page
export type NavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  'Login'
>;

export type ToastRenderProps = {
  id: string;
};

export type UserItem = {
  email: string;
  familyName: string;
  givenName: string;
  id: string;
  name: string;
  photo: string;
};

export type HandleBookCommentFunction = {
  handleNavigateUserNotLogin: () => void;
  handleAddCommentBook: (review: AddCommentValues) => void;
  handleDeleteCommentBook: () => void;
  handleUpdateCommentBook: (ulasan: string, updateAt: string) => void;
  getIdDocumentFromFirestore: (id: string) => void;
};

export type CommentProps = {
  refOpenDeleteModal: React.RefObject<any>;
  deleteModal: boolean;
  setDeleteModal: React.Dispatch<React.SetStateAction<boolean>>;
  handleButton: HandleBookCommentFunction;
  isEdit: boolean;
  setIsEdit: React.Dispatch<React.SetStateAction<boolean>>;
  commentSchema: ObjectSchema<FormCommentValues>;
  initialValues: FormCommentValues;
  isLogin: boolean;
  user: UserItem;
  detailBook: BookItem;
  commentBook: Array<Comment>;
};
