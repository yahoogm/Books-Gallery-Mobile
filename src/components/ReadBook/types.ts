export type ReadBookProps = {
  setShowBook: React.Dispatch<React.SetStateAction<boolean>>;
  showBook: boolean;
  refReadBook: React.RefObject<any>;
  isbn: number;
};
