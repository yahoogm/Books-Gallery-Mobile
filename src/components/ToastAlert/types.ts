export type ToastAlertProps = {
  id: string;
  message: string;
  action: 'success' | 'info' | 'error' | 'warning' | 'attention';
};
