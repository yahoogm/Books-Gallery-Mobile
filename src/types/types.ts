type VolumeInfo = {
  title: string;
  publishedDate: string;
  description: string;
  authors: Array<string>;
  publisher: string;
  pageCount: number;
  imageLinks: {smallThumbnail: string; thumbnail: string; large?: string};
  categories: Array<string>;
  industryIdentifiers: Array<{identifier: string; type: string}>;
};

export type Review = {
  bookId: string;
  createdAt: {};
  id: string;
  profilePic: string;
  ulasan: string;
  updatedAt: {};
  userId: string;
  userName: string;
};

export type BookItem = {
  id: string;
  volumeInfo: VolumeInfo;
};

export type BookState = {
  type: string;
  reviewId: string;
  detailBook: BookItem;
  books: {
    items: [];
  };
  reviewBook: Array<Review>;
};

export type userItem = {
  email: string;
  familyName: string;
  givenName: string;
  id: string;
  name: string;
  photo: string;
};

export type RootStackParamList = {
  DetailBook: {bookId: string};
  Login: undefined;
};
