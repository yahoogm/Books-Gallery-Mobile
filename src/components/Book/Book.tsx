import {Button, ButtonText, View, Image} from '@gluestack-ui/themed';
import {BookProps} from './types';
import BookItem from '../BookItem/BookItem';

const Book: React.FC<BookProps> = ({
  uriDetailBook,
  authorsName,
  categories,
  detailBook,
  setShowBook,
  refReadBook,
}) => {
  return (
    <View display="flex" alignItems="center" padding={20}>
      <Image
        alt="title"
        width={250}
        h={340}
        borderRadius={10}
        source={{
          uri: uriDetailBook,
        }}
      />

      <View display="flex" flexDirection="row" gap={10} marginTop={10}>
        <Button onPress={() => setShowBook(true)} ref={refReadBook}>
          <ButtonText>Baca Buku</ButtonText>
        </Button>
      </View>

      <View alignItems="center" marginTop={10}>
        <BookItem
          detailBook={
            authorsName !== null || undefined
              ? authorsName
              : 'Penulis tidak diketahui'
          }
          label="Author"
        />
        <BookItem
          detailBook={
            detailBook.volumeInfo.publisher !== null || undefined
              ? detailBook.volumeInfo.publisher
              : 'Penerbit tidak diketahui'
          }
          label="Publisher"
        />
        <BookItem
          detailBook={
            detailBook.volumeInfo.publishedDate !== null || undefined
              ? detailBook.volumeInfo.publishedDate
              : 'Waktu diterbitkan tidak diketahui'
          }
          label="Published Date"
        />
        <BookItem
          detailBook={
            detailBook.volumeInfo.pageCount !== null || undefined
              ? detailBook.volumeInfo.pageCount
              : 'Jumlah halaman tidak diketahui'
          }
          label="Page Count"
        />
        <BookItem
          detailBook={
            categories !== null || undefined
              ? categories
              : 'Kategori tidak diketahui'
          }
          label="Category"
        />
        <BookItem
          detailBook={
            detailBook.volumeInfo.description !== null || undefined
              ? detailBook.volumeInfo.description
              : 'Deskripsi tidak diketahui'
          }
          label="Description"
        />
        {/* <View alignItems="center">
          <Heading size="md">Author:</Heading>
          <Text>
            {authorsName !== null ? authorsName : 'Penulis tidak diketahui'}
          </Text>
        </View>
        <View alignItems="center">
          <Heading size="md">Publisher:</Heading>
          <Text>
            {detailBook.volumeInfo.publisher !== null || undefined
              ? detailBook.volumeInfo.publisher
              : 'Penerbit tidak diketahui'}
          </Text>
        </View>
        <View alignItems="center">
          <Heading size="md">Published Date:</Heading>
          <Text>
            {detailBook.volumeInfo.publishedDate !== null || undefined
              ? detailBook.volumeInfo.publishedDate
              : 'Tanggal publikasi tidak diketahui'}
          </Text>
        </View>
        <View alignItems="center">
          <Heading size="md">Page Count:</Heading>
          <Text>
            {detailBook.volumeInfo.pageCount !== null || undefined
              ? detailBook.volumeInfo.pageCount
              : 'Jumlah halaman tidak diketahui'}
          </Text>
        </View>
        <View alignItems="center">
          <Heading size="md">Category:</Heading>
          <Text>
            {(categories !== undefined && categories.length !== 0) ||
            null ||
            undefined
              ? categories
              : 'Kategori tidak diketahui'}
          </Text>
        </View>
        <View alignItems="center">
          <Heading size="md">Description:</Heading>
          <Text>
            {detailBook.volumeInfo.description !== null || undefined
              ? detailBook.volumeInfo.description
              : 'Deksripsi tidak diketahui'}
          </Text>
        </View> */}
      </View>
    </View>
  );
};

export default Book;
