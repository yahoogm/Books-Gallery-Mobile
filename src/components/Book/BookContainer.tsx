import Book from './Book';
import {useAppSelector} from '../../hooks/useRedux';
import {detailBookSelector} from '../../redux/book/bookSelector';
import {modifiedName} from '../../../utils/const/const';
import {BookContainerProps} from './types';

const BookContainer: React.FC<BookContainerProps> = ({
  refReadBook,
  setShowBook,
}) => {
  const detailBook = useAppSelector(detailBookSelector);

  const authorsName = modifiedName(detailBook.volumeInfo.authors);
  const categories = modifiedName(detailBook.volumeInfo.categories);

  const uriDetailBook =
    detailBook.volumeInfo.imageLinks &&
    detailBook.volumeInfo.imageLinks.thumbnail
      ? detailBook.volumeInfo.imageLinks.thumbnail
      : 'https://plus.unsplash.com/premium_photo-1698084059435-a50ddfd69303?q=80&w=1850&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D';
  return (
    <Book
      setShowBook={setShowBook}
      refReadBook={refReadBook}
      detailBook={detailBook}
      categories={categories}
      uriDetailBook={uriDetailBook}
      authorsName={authorsName}
    />
  );
};

export default BookContainer;
