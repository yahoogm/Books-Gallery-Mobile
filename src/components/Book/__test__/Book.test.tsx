import React, {createRef} from 'react';
import {render} from '@testing-library/react-native';
import {Button} from '@gluestack-ui/themed';
import Book from '../Book';
import {BookProps} from '../types';
import {modifiedName} from '../../../utils/const/const';

const mockDetailBook = {
  volumeInfo: {
    title: 'Test Title',
    publishedDate: '2024-01-01',
    description: 'Test Description',
    authors: ['Author One', 'Author Two'],
    publisher: 'Test Publisher',
    pageCount: 123,
    imageLinks: {
      smallThumbnail: 'http://example.com/small.jpg',
      thumbnail: 'http://example.com/thumbnail.jpg',
    },
    categories: ['Category One', 'Category Two'],
    industryIdentifiers: [
      {identifier: '1234567890', type: 'ISBN_10'},
      {identifier: '1234567890123', type: 'ISBN_13'},
    ],
  },
};

describe('BookComponent', () => {
  const rendererComponent = (props: BookProps) => render(<Book {...props} />);

  // snapshot test
  it('match snapshot when given props', () => {
    const mockSetShowBook = jest.fn();
    const refReadBook = createRef<any>();
    const props = {
      authorsName: ['budiono siregar', 'ucup'],
      categories: ['buku', 'catatan'],
      detailBook: mockDetailBook,
      refReadBook: refReadBook,
      setShowBook: mockSetShowBook,
      uriDetailBook: 'https://google.com/test',
    };

    const tree = rendererComponent(props).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // assertion test
  it('calls setShowBook when button pressed', () => {
    const mockSetShowBook = jest.fn();
    const refReadBook = createRef<any>();

    const props = {
      authorsName: ['budiono siregar', 'ucup'],
      categories: ['buku', 'catatan'],
      detailBook: mockDetailBook,
      refReadBook: refReadBook,
      setShowBook: mockSetShowBook,
      uriDetailBook: 'https://google.com/test',
    };

    const tree = rendererComponent(props);
    const testInstance = tree.root;
    const button = testInstance.findByType(Button);
    button.props.onPress();

    expect(mockSetShowBook).toHaveBeenCalledWith(true);
    expect(refReadBook.current).toBeTruthy();
  });

  it('renders correctry when given props', () => {
    const mockSetShowBook = jest.fn();
    const refReadBook = createRef<any>();
    const names = modifiedName(['budiono siregar', 'ucup']);
    const categories = modifiedName(['buku', 'catatan']);

    const props = {
      authorsName: names,
      categories: categories,
      detailBook: mockDetailBook,
      refReadBook: refReadBook,
      setShowBook: mockSetShowBook,
      uriDetailBook: 'https://google.com/test',
    };

    const tree = rendererComponent(props);
    const instance = tree.root;

    expect(
      instance.findByProps({children: 'budiono siregar, ucup'}),
    ).toBeTruthy();
    expect(instance.findByProps({children: 'buku, catatan'})).toBeTruthy();
  });
});
