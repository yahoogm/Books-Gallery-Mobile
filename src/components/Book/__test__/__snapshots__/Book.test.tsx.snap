// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookComponent match snapshot when given props 1`] = `
<View
  data-style=""
  dataSet={
    {
      "style": "",
    }
  }
  style={
    [
      undefined,
      {},
      {
        "alignItems": "center",
        "display": "flex",
        "padding": 20,
      },
      undefined,
    ]
  }
>
  <Image
    alt="title"
    aria-label="title"
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    h={340}
    role="img"
    source={
      {
        "uri": "https://google.com/test",
      }
    }
    style={
      [
        undefined,
        {},
        {
          "borderRadius": 10,
          "width": 250,
        },
        undefined,
      ]
    }
  />
  <View
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        {
          "display": "flex",
          "flexDirection": "row",
          "gap": 10,
          "marginTop": 10,
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      data-style=""
      dataSet={
        {
          "active": "false",
          "disabled": "false",
          "focus": "false",
          "focusVisible": "false",
          "hover": "false",
          "style": "",
        }
      }
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      role="button"
      style={
        [
          undefined,
          {},
          undefined,
        ]
      }
    >
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
            {},
            undefined,
          ]
        }
      >
        Baca Buku
      </Text>
    </View>
  </View>
  <View
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        {
          "alignItems": "center",
          "marginTop": 10,
        },
        undefined,
      ]
    }
  >
    <View
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        accessibilityRole="header"
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginVertical": 42.56,
          }
        }
      >
        Author:
      </Text>
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
          ]
        }
      >
        budiono siregar
        ucup
      </Text>
    </View>
    <View
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        accessibilityRole="header"
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginVertical": 42.56,
          }
        }
      >
        Publisher:
      </Text>
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
          ]
        }
      >
        Test Publisher
      </Text>
    </View>
    <View
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        accessibilityRole="header"
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginVertical": 42.56,
          }
        }
      >
        Published Date:
      </Text>
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
          ]
        }
      >
        2024-01-01
      </Text>
    </View>
    <View
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        accessibilityRole="header"
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginVertical": 42.56,
          }
        }
      >
        Page Count:
      </Text>
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
          ]
        }
      >
        123
      </Text>
    </View>
    <View
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        accessibilityRole="header"
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginVertical": 42.56,
          }
        }
      >
        Category:
      </Text>
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
          ]
        }
      >
        buku
        catatan
      </Text>
    </View>
    <View
      data-style=""
      dataSet={
        {
          "style": "",
        }
      }
      style={
        [
          undefined,
          {},
          {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <Text
        accessibilityRole="header"
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        size="md"
        style={
          {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginVertical": 42.56,
          }
        }
      >
        Description:
      </Text>
      <Text
        data-style=""
        dataSet={
          {
            "style": "",
          }
        }
        style={
          [
            undefined,
            {},
            undefined,
          ]
        }
      >
        Test Description
      </Text>
    </View>
  </View>
</View>
`;
